### PR TITLE
[wip] Add SyftTensor, FloatTensor, DataTensor, IntegerTensor

### PR DESCRIPTION
**Pull request by koenvanderveen**
## Description
aims to implement #5046

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by madhavajay** [comment]
@koenvanderveen great start! ❤️

File: src/syft/lib/tensor/tensorbase.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+Num = Union[int, float]
+
+
+class ChildDelegatorTensor:
+    wrap_type: Any = None
+
+    def __getattr__(self, name: str) -> Any:
+        func_or_attr = getattr(self.child, name)
+        if isfunc(func_or_attr):
```
Comment (madhavajay) [inline src/syft/lib/tensor/tensorbase.py:24]: Would `callable` replace isfunc? Also is there any need to some how create a dynamic wrapper for @property or `getset_descriptor` attributes?

**Review by madhavajay — APPROVED** [review]
This is awesome. 👍🏼

File: src/syft/lib/tensor/tensorbase.py (hunk 1 above)
Comment (koenvanderveen) [reply src/syft/lib/tensor/tensorbase.py]: updated

File: src/syft/lib/tensor/tensorbase.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        if isinstance(self.child, FloatTensor) and isinstance(other, int):
+            return SyftTensor(child=self.child / other)
+        if isinstance(self.child, IntegerTensor) and isinstance(other, float):
+            return SyftTensor(child=self.child / other)
+        else:
+            raise ValueError()
+
+    @syft_decorator(typechecking=True, prohibit_args=False)
+    def __matmul__(self, other: "SyftTensor") -> "SyftTensor":
+        if isinstance(self.child, FloatTensor) and isinstance(other, SyftTensor):
```
Comment (gmuraru) [inline src/syft/lib/tensor/tensorbase.py:158]: Q: Why not use?
```
if isinstance(self.child, (FloatTensor, IntegerTensor))
```
In that case there will be only 1 ```if``` statement

File: src/syft/lib/tensor/tensorbase.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        func_or_attr = getattr(self.child, name)
+        if callable(func_or_attr):
+            return partial(call_func_and_wrap_result, func_or_attr, self.__class__)
+        else:
+            var = func_or_attr
+            return var
+
+
+class DataTensor(ChildDelegatorTensor):
+    @syft_decorator(typechecking=True)
```
Comment (gmuraru) [inline src/syft/lib/tensor/tensorbase.py:29]: Q: @tudorcebere we should remove the ```syft_decorator```, right?

File: src/syft/lib/tensor/tensorbase.py (hunk 2 above)
Comment (koenvanderveen) [reply src/syft/lib/tensor/tensorbase.py:158]: Oh yes forgot about that functionality

Comment (gmuraru) [reply src/syft/lib/tensor/tensorbase.py:158]: Sorry for not giving more information!

I meant this if and the next one.
``` 
if isinstance(self.child, FloatTensor) and isinstance(other, SyftTensor):
  return SyftTensor(child=self.child @ other.child)
if isinstance(self.child, IntegerTensor) and isinstance(other, SyftTensor):
  return SyftTensor(child=self.child @ other.child)
```

Changed to:
``` 
if isinstance(self.child, (FloatTensor, IntegerTensor)) and isinstance(other, SyftTensor):
  return SyftTensor(child=self.child @ other.child)
```

File: src/syft/lib/tensor/tensorbase.py (hunk 3 above)
Comment (gmuraru) [reply src/syft/lib/tensor/tensorbase.py:29]: Don't forget about this :D

File: src/syft/lib/tensor/tensorbase.py (hunk 2 above)
Comment (madhavajay) [reply src/syft/lib/tensor/tensorbase.py]: Turns out this isn't possible due to Mypy, I get the error:
```
Both left and right operands are unions
```

I tried creating a Union type but isinstance can't work with "Typing" stuff. There could be a way around this but for now I think its okay.

File: src/syft/lib/tensor/tensorbase.py (hunk 3 above)
Comment (madhavajay) [reply src/syft/lib/tensor/tensorbase.py]: I removed all of these. 😊

**Pull request merged**